test/e2e/suite/command/custom_header.go

```diff
@@ -87,6 +87,22 @@ var _ = Describe("Common registry users:", func() {
 				"-H", FoobarHeaderInput, "-H", AbHeaderInput).
 				MatchRequestHeaders(FoobarHeader, AbHeader).Exec()
 		})
+		It("copy and add custom headers in source registry requests", func() {
+			headerTestRepo := headerTestRepo("from-header")
+			fromUrlPrefix := fmt.Sprintf("://%s/v2/%s/", Host, ImageRepo)
+			src := RegistryRef(Host, ImageRepo, foobar.Tag)
+			dst := RegistryRef(Host, headerTestRepo, "fromHeader")
+			ORAS("cp", src, dst, "-d", "--from-header", FoobarHeaderInput, "--from-header", AbHeaderInput).
+				MatchCpRequestHeaders(fromUrlPrefix, FoobarHeader, AbHeader).Exec()
+		})
+		It("copy and add custom headers in destination registry requests", func() {
+			headerTestRepo := headerTestRepo("to-header")
+			toUrlPrefix := fmt.Sprintf("://%s/v2/%s/", Host, headerTestRepo)
```

Comment:
please check comments on `NewRequestHeaderMatcherWithPrefix` and move this prefix-assembling snippet there.